pkg/provider/core.go

```diff
@@ -101,13 +103,22 @@ func (p *Provider) CreateMachine(ctx context.Context, req *driver.CreateMachineR
 	}
 	instanceMetadata["user-data"] = string(userData)
 
+	folderId := string(secret.Data[api.YandexFolderID])
+
 	var networkInterfaceSpecs []*compute.NetworkInterfaceSpec
 	for _, netIf := range spec.NetworkInterfaceSpecs {
 		var natSpec *compute.OneToOneNatSpec
 		if netIf.AssignPublicIPAddress {
 			natSpec = &compute.OneToOneNatSpec{
 				IpVersion: compute.IpVersion_IPV4,
 			}
+
+			if len(netIf.PublicIPAddresses) > 0 {
+				natSpec.Address, err = findUnusedAddressFromListByFolder(ctx, ySDK, folderId, netIf.PublicIPAddresses)
+				if err != nil {
+					return nil, mcmstatus.Error(mcmcodes.InvalidArgument, err.Error())
+				}
```

Comment:
we can use [`NotFound`](https://github.com/gardener/machine-controller-manager/blob/78057547c9bb7208ca254943f0f263c16d661912/pkg/util/provider/machinecodes/codes/codes.go#L59) code , here as we can't consider the argument as invalid ( no spelling mistakes or wrong type)